YoCode/ProcessRunner.cs

```diff
@@ -2,43 +2,93 @@
-        public bool TimedOut { get; set; }
-        ProcessInfo procinfo;
+        public bool TimedOut { get; private set; }
         public string Output { get; set; }
         public string ErrorOutput { get; set; }
 
-        public ProcessRunner(string processName,string workingDir,string arguments)
+        private ProcessInfo procinfo;
+        private readonly TimeSpan timeout = TimeSpan.FromSeconds(15);
+        private readonly StringBuilder output = new StringBuilder();
+        private readonly StringBuilder errorOutput = new StringBuilder();
+
+
+        public ProcessRunner(string processName, string workingDir, string arguments)
         {
             procinfo = SetupProcessInfo(processName, workingDir, arguments);
         }
 
-        public void ExecuteTheCheck()
+        public void ExecuteTheCheck(string waitForMessage = null)
         {
             var p = new Process();
             p.StartInfo = SetProcessStartInfo(procinfo);
+            p.EnableRaisingEvents = true;
+            p.OutputDataReceived += DataReceived;
+            p.ErrorDataReceived += ErrorDataReceived;
             p.Start();
+            p.BeginOutputReadLine();
+            p.BeginErrorReadLine();
+
+            WaitForExitCondition(p, waitForMessage);
+
+            Output = output.ToString();
+            ErrorOutput = errorOutput.ToString();
+        }
 
-            if(!p.WaitForExit((int)timeout.TotalMilliseconds))
+        private void WaitForExitCondition(Process p, string wait)
+        {
+            if (string.IsNullOrEmpty(wait))
             {
-                TimedOut = true;
+                if (!p.WaitForExit((int)timeout.TotalMilliseconds))
+                {
+                    TimedOut = true;
+                }
+                return;
             }
-            Output = p.StandardOutput.ReadToEnd();
-            ErrorOutput = p.StandardError.ReadToEnd();
+
+            /*If code gets to this point, it should also look for "Unable to start Kestrel"
+             *in case kestrel fails (Fails to bind to address)*/
+            List<string> keywords = new List<string>
+            {
+                wait,
+                "Unable to start Kestrel"
+            };
+
+            /*Check if 'appsettings.json file is missing*/
+            List<string> errorKeywords = new List<string>
+            {
+                "'appsettings.json' was not found and is not optional."
+            };
+
+            var loopRetryDelay = TimeSpan.FromSeconds(0.5);
+            var numberOfTimesToRetry = (int)(timeout / loopRetryDelay);
+            int numberOfRetries = 0;
+
+            while (!p.HasExited
+                && !output.ToString().ContainsAny(keywords)
+                && numberOfRetries < numberOfTimesToRetry
+                && !errorOutput.ToString().ContainsAny(errorKeywords))
+            {
+                Thread.Sleep(loopRetryDelay);
+                numberOfRetries++;
+            }
+
+            try { p.Kill(); }
```

Comment:
I wonder if this might be nicer as
```
if (!p.HasExited)
{
    p.Kill();
}
```
It should mean we can avoid the try/catch and the need for a comment to explain.